README

                          uncompyle2 
        A Python 2.5, 2.6, 2.7 byte-code decompiler, written in Python 2.7
                          0.13
                          2012-6-5

Introduction
------------

'uncompyle2' converts Python byte-code back into equivalent Python
source. It accepts byte-code from Python version 2.5, 2.6 & 2.7. Additionally,
it will only run on Python 2.7.

The generated source is very readable: docstrings, lists, tuples and
hashes get pretty-printed.

'uncompyle2' may also verify the equivalence of the generated source by
by compiling it and comparing both byte-codes.

'uncompyle2' is based on John Aycock's generic small languages compiler
'spark' (http://www.csr.uvic.ca/~aycock/python/) and his prior work on
'uncompyle'.

Additional note (3 July 2004, Ben Burton):

    The original website from which this software was obtained is no longer
    available.  It has now become a commercial decompilation service, with
    no software available for download.

    Any developers seeking to make alterations or enhancements to this code
    should therefore consider these debian packages an appropriate starting
    point.

Additional note (5 June 2012):

    The decompilation of python bytecode 2.5 & 2.6 is based on the work of
    Eloi Vanderbeken. bytecode is translated to a pseudo 2.7 python bytecode
    and then decompiled.

Features
--------

  * decompiles Python byte-code into equivalent Python source

  * decompiles byte-code from Python version 2.5, 2.6, 2.7

  * pretty-prints docstrings, hashes, lists and tuples
  
  * reads directly from .pyc/.pyo files, bulk-decompile whole
    directories

  * output may be written to file, a directory or to stdout

  * option for including byte-code disassembly into generated source

  For a list of changes please refer to the 'CHANGES' file.


Requirements
------------

uncompyle2 requires Python 2.7


Installation
------------

You may either create a RPM and install this, or install directly from
the source distribution.

Creating RPMS:

  python setup.py bdist_rpm

  If you need to force the python interpreter to eg. pyton2:
     python2 setup.py bdist_rpm --python=python2


Installation from the source distribution:

     python setup.py install

   To install to a user's home-dir:
     python setup.py install --home=<dir>

   To install to another prefix (eg. /usr/local)
     python setup.py install --prefix=/usr/local

  If you need to force the python interpreter to eg. pyton2:
     python2 setup.py install

  For more information on 'Installing Python Modules' please refer to
     http://www.python.org/doc/current/inst/inst.html


Usage
-----

uncompyle2 -h		prints short usage
uncompyle2 --help	prints long usage


Known Bugs/Restrictions
-----------------------

I have some known bug in the 2.5 & 2.6 decompilation version but this will be fixed in a few.


NEWS
uncompyle6 3.2.5 2019-12-30 Clearout sale

- 3.7.2 Remove deprecation warning on regexp string that isn't raw
- main.main() parameter `codes` is not used - note that
- Improve Python 3.6+ control flow detection
- More complete fragment instruction annotation for `imports`

uncompyle6 3.2.4 2018-10-27 7x9 release

- Bug fixes #180, #182, #187, #192
- Enhancements #189
- Internal improvements

uncompyle6 3.2.3 2018-06-04 Michael Cohen flips and Fleetwood Redux

- Python 1.3 support 3.0 bug and
- fix botched parameter ordering of 3.x in last release

uncompyle6 3.2.2 2018-06-04 When I'm 64

- Python 3.0 support and bug fixes

uncompyle6 3.2.1 2018-06-04 MF

- Python 1.4 and 1.5 bug fixes

uncompyle6 3.2.0 2018-05-19 Rocket Scientist

- Add rudimentary 1.4 support (still a bit buggy)
- add --tree+ option to show formatting rule, when it is constant
- Python 2.7.15candidate1 support (via xdis)
- bug fixes, especially for 3.7 (but 2.7 and 3.6 and others as well)

uncompyle6 3.1.3 2018-04-16

- Add some Python 3.7 rules, such as for handling LOAD_METHOD (not complete)
- Fix some fragment bugs
- small doc changes

uncompyle6 3.1.2 2018-04-08 Eastern Orthodox Easter

- Python 3.x subclass and call parsing fixes
- Allow/note running on Python 3.1
- improve 3.5+ BUILD_MAP_UNPACK
- DRY instruction building code between 2.x and 3.x
- expand testing

uncompyle6 3.1.1 2018-04-01 Easter April Fool's

Jesus on Friday's New York Times puzzle: "I'm stuck on 2A"

- fill out 3.5+ BUILD_MAP_UNPACK (more work is needed)
- fill out 3.4+ CALL_FUNCTION_... (more work is needed)
- fill out 3.5 MAKE_FUNCTION  (more work is needed)
- reduce 3.5, 3.6 control-flow bugs
- reduce ambiguity in rules that lead to long (exponential?) parses
- limit/isolate some 2.6/2.7,3.x grammar rules
- more runtime testing of decompiled code
- more removal of parenthesis around calls via setting precidence

uncompyle6 3.1.0 2018-03-21 Equinox

- Add code_deparse_with_offset() fragment function.
- Correct paramenter call fragment deparse_code()
- Lots of 3.6, 3.x, and 2.7 bug fixes
  About 5% of 3.6 fail parsing now. But
  semantics still needs much to be desired.

uncompyle6 3.0.1 2018-02-17

- All Python 2.6.9 standard library files weakly verify
- Many 3.6 fixes. 84% of the first 200 standard library files weakly compile.
  One more big push is needed to get the remaining to compile
- Many decompilation fixes for other Python versions
- Add more to the test framework
- And more add tests target previous existing bugs more completely
- sync recent license changes in metadata

uncompyle6 3.0.0 2018-02-17

- deparse_code() and lookalikes from the various semantic actions are
  now deprecated. Instead use new API code_deparse() which makes the
  version optional and bundles debug options into a dictionary.
- License changed to GPL3.
- Many Python 3.6 fixes, especially around handling EXTENDED_ARGS
  Due to the reduction in operand size for JUMP's there are many
  more EXTENDED_ARGS instructions which can be the targets
  of jumps, and messes up the peephole-like analysis that is
  done for control flow since we don't have something better in place.
- Code has been reorganized to be more instruction nametuple based where it
  has been more bytecode array based. There was and still is code that had
  had magic numbers to advance instructions or to pick out operands.
- Bug fixes in numerous other Python versions
- Instruction display improved
- Keep global statements in fixed order (from wangym5106)

A bit more work is still needed for 3.6 especially in the area of
function calls and definitions.


uncompyle6 2.16.0 2018-02-17

- API additions:
  - add fragments.op_at_code_loc() and
  - fragments.deparsed_find()_
- Better 2.7 end_if and COME_FROM determination
- Fix up 3.6+ CALL_FUNCTION_EX
- Misc pydisasm fixes
- Weird comprehension bug seen via new loctraceback
- Fix Python 3.5+ CALL_FUNCTION_VAR and BUILD_LIST_UNPACK in call; with this
  we can can handle 3.5+ f(a, b, *c, *d, *e) now

uncompyle6 2.15.1 2018-02-05

- More bug fixes and revert an improper bug fix in 2.15.0

uncompyle6 2.15.0 2018-02-05 pycon2018.co

- Bug fixes
- Code fragment improvements
- Code cleanups
- Expand testing

uncompyle6 2.15.1 2018-01-27

- Add --linemap option to give line correspondences
  between original source lines and reconstructed line sources.
  It is far from perfect, but it is a start
- Add a new class of tests: tests which when decompiled check themselves
- Split off Python version semantic action customizations into its own file
- Fix 2.7 bug in ifelse loop statement
- Handle 3.6+ EXTENDED_ARGs for POP_JUMP_IF... instructions
- Correct 3.6+ calls with kwargs
- Describe the difficulty of 3.6 in README

uncompyle6 2.14.3 2018-01-19

- Fix bug in 3.5+ await stmt
- Better version to magic handling; handle 3.5.2 .. 3.5.4 versions
- Improve/correct test_pyenvlib.py status messages
- Fix some 2.7 and 2.6 parser bugs
- Fix whilelse parsing bugs
- Correct 2.5- decorator parsing
- grammar for decorators matches AST a little more
- better tests in setup.py for running the right version of Python
- Fix 2.6- parsing of "for .. try/else" ...  with "continue"  inside

uncompyle6 2.14.2 2018-01-09 Samish

Decompilation bug fixes, mostly 3.6 and pre 2.7

- 3.6 FUNCTION_EX (somewhat)
- 3.6 FUNCTION_EX_KW fixes
- 3.6 MAKE_FUNCTION fixes
- correct 3.5 CALL_FUNCTION_VAR
- stronger 3.x "while 1" testing
- Fix bug in if's with "pass" bodies. Fixes #104
- try/else and try/finally fixes on 2.6-
- limit pypy customization to pypy
- Add addr fields in COME_FROMS
- Allow use of full instructions in parser reduction routines
- Reduce grammar in Python 3 by specialization more to specific
  Python versions
- Match Python AST names more closely when possible

uncompyle6 2.14.1 2017-12-10 Dr. Gecko

- Many decompilation bugfixes
- Grammar rule reduction and version isolation
- Match higher-level nonterminal names more closely
  with Python AST

__pkginfo__.py
# Copyright (C) 2018 Rocky Bernstein <rocky@gnu.org>
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""uncompyle6 packaging information"""

# To the extent possible we make this file look more like a
# configuration file rather than code like setup.py. I find putting
# configuration stuff in the middle of a function call in setup.py,
# which for example requires commas in between parameters, is a little
# less elegant than having it here with reduced code, albeit there
# still is some room for improvement.

# Things that change more often go here.
copyright   = """
Copyright (C) 2015-2018 Rocky Bernstein <rb@dustyfeet.com>.
"""

classifiers =  ['Development Status :: 5 - Production/Stable',
                'Intended Audience :: Developers',
                'License :: OSI Approved :: GNU General Public License v3 (GPLv3)',
                'Operating System :: OS Independent',
                'Programming Language :: Python',
                'Programming Language :: Python :: 2.4',
                'Programming Language :: Python :: 2.5',
                'Programming Language :: Python :: 2.6',
                'Programming Language :: Python :: 2.7',
                'Programming Language :: Python :: 3.0',
                'Programming Language :: Python :: 3.1',
                'Programming Language :: Python :: 3.2',
                'Programming Language :: Python :: 3.3',
                'Programming Language :: Python :: 3.4',
                'Programming Language :: Python :: 3.5',
                'Programming Language :: Python :: 3.6',
                'Programming Language :: Python :: 3.7',
                'Topic :: Software Development :: Debuggers',
                'Topic :: Software Development :: Libraries :: Python Modules',
                ]

# The rest in alphabetic order
author             = "Rocky Bernstein, Hartmut Goebel, John Aycock, and others"
author_email       = "rb@dustyfeet.com"
entry_points       = {
    'console_scripts': [
        'uncompyle6=uncompyle6.bin.uncompile:main_bin',
        'pydisassemble=uncompyle6.bin.pydisassemble:main',
    ]}
ftp_url            = None
install_requires   = ['spark-parser >= 1.8.7, < 1.9.0',
                      'xdis >= 3.8.9, < 3.9.0']

license            = 'GPL3'
mailing_list       = 'python-debugger@googlegroups.com'
modname            = 'uncompyle6'
py_modules         = None
short_desc         = 'Python cross-version byte-code decompiler'
web                = 'https://github.com/rocky/python-uncompyle6/'

# tracebacks in zip files are funky and not debuggable
zip_safe = True


import os.path
def get_srcdir():
    filename = os.path.normcase(os.path.dirname(os.path.abspath(__file__)))
    return os.path.realpath(filename)

srcdir = get_srcdir()

def read(*rnames):
    return open(os.path.join(srcdir, *rnames)).read()

# Get info from files; set: long_description and VERSION
long_description   = ( read("README.rst") + '\n' )
exec(read('uncompyle6/version.py'))
